added: docs for kyverno metrics

yashvardhan-kukreja opened the pull request:
Signed-off-by: Yashvardhan Kukreja <yash.kukreja.98@gmail.com>

Added docs around the upcoming Prometheus metrics integration of Kyverno.

Check this out for further reference: https://github.com/kyverno/kyverno/pull/1877

JimBugwadia commented:
preview: https://deploy-preview-170--kyverno.netlify.app/docs/monitoring-kyverno-with-prometheus-metrics/

chipzoller commented:
Rather than the folder/sidebar entry being "Monitoring Kyverno With Prometheus Metrics" I recommend that be shortened to simply "Monitoring Kyverno"

JimBugwadia commented:
Yes, agree on shortening titles. Same for:

`Bonus: A ready-to-use Grafana dashboard depicting insights from Kyverno metrics`

Could be shortened to `Grafana dashboard` and the subtitle could be `A ready to use dashboard depicting insights from Kyverno metrics`.

JimBugwadia commented:
For the metrics we can come up with shorter names that explain what each metric provides:

`kyverno_admission_review_latency_milliseconds` -> `API Request Latency`

etc.
etc.

JimBugwadia commented:
I was also wondering if the chapter is better placed after `Policy Reports`.

Any thoughts?

yashvardhan-kukreja commented:
Thanks @JimBugwadia for your suggestions. I absolutely agree with all of them except one:

> For the metrics we can come up with shorter names that explain what each metric provides:
> 
> `kyverno_admission_review_latency_milliseconds` -> `API Request Latency`
> 
> etc.
> 
> 

Actually `kyverno_admission_review_latency_milliseconds` was named such way to comply with the Prometheus's established naming convention i.e  the metric should begin with the project/tool's name (kyverno), then should lead upto what the metric is serving(admission_review_latency) and end with the unit (milliseconds) in which it is serving the metric.

So, taking your suggestion, the metric would be renamed to `kyverno_api_request_latency_milliseconds`but that is pretty much similar to the length of `kyverno_admission_review_latency_milliseconds`.

Also, thinking as a user, it seems hard to decode the meaning of API request latency and correlate it to the actual meaning of it, which is the latency of the admission request/review.

Hence, as a user, `kyverno_admission_review_latency_milliseconds` sounds more understandable to me in the sense that we are talking about the latency of the reviews of entire admission requests here.

Also, metric names of this length are very normal as per my experience with multiple (cloud/kubernetes)-native tools. For example, kube-state-metrics.

So, I'd suggest to keep this metric's name the same only :)

chipzoller reviewed (changes requested):
See above items.

yashvardhan-kukreja commented:
> I was also wondering if the chapter is better placed after `Policy Reports`.
> 
> Any thoughts?

I feel that the docs around Monitoring Kyverno should have it's own section independent from `Policy Reports` (as provided in the current state in the PR) because, although both of them might be a intersecting a bit, still both of them don't seem to be majorly correlated to each other  from a user-facing standpoint in the sense that one is exposing a wide-variety of metrics ranging from admission requests to background scans to rule-level performance, in a Prometheus-readable format and the other is just capturing results around Policies' executions and storing them as a Kubernetes-resource. 
Also, `Monitoring Kyverno`'s docs are anyways pretty nested so bringing them as a child of the `Policy Reports` would end up just adding another layer of nesting in the docs which seems a bit unclean to me.

JimBugwadia commented:
Hi @yashvardhan-kukreja 

> So, I'd suggest to keep this metric's name the same only :)

I was suggesting using simpler names for the document header titles, not suggesting we need to change the underlying metric names. We can use "Admission Review Latency" as the title to align. Then, in the document text, you can provide the full metric name.


> I feel that the docs around Monitoring Kyverno should have it's own section independent from Policy Reports...

Yes, agree it should be its own section.  I was suggesting to place that section **after** the Policy Reports section in the index.

yashvardhan-kukreja commented:
Ohh haha, my bad, I misread. Sure Jim, I'll implement the above changes.
Although seemingly, the hugo theme automatically rearranges the index alphabetically, so, that's why Monitoring Kyverno is coming before Policy Reports.
But no worries, let me try to figure out a way to deal with this alphabetic arrangement :)

JimBugwadia commented on content/en/docs/Monitoring Kyverno With Prometheus Metrics/_index.md:
```diff
@@ -0,0 +1,55 @@
+  ## Kyverno's metrics server will be exposed at this port
+  metricsPort: 8000
+  ## The Node's port which will allow access Kyverno's metrics at the host level. Only used if service.type is NodePort.
+  metricsNodePort: 8000
```
Lets remove this, as Kyverno does not use a node port

yashvardhan-kukreja replied:
So, the helm chart is configured in such a way that Kyverno can expose itself as a clusterIP (default and private), NodePort (publicly exposed via a host's port) and LoadBalancer (most generically used - publicly exposed via a cloud-provisioned LoadBalancer).
For more details, please checkout the following comment. It has some important details around this.

JimBugwadia replied:
Ok, got it.

JimBugwadia commented on content/en/docs/Monitoring Kyverno With Prometheus Metrics/_index.md:
````diff
@@ -0,0 +1,55 @@
+...
+```
+
+The above configuration will allow Kyverno's metrics to be exposed at port 8000 of any host/node of the cluster.
````
I would suggest we simplify this (line 38 - 53) to:

By default, Kyverno installs with service type `ClusterIP`. To expose the Kyverno service outside the cluster you can configure a different [service type](https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types), or use an Ingress (https://kubernetes.io/docs/concepts/services-networking/ingress/0/).

yashvardhan-kukreja replied:
So Jim, the way this helm chart is configured, it allows the user to expose the kyverno service of any type i.e. `ClusterIP` (default and only privately accessibly), `NodePort` (publicly accessible over a host's port), `LoadBalancer` (publicly accessible over a cloud-provisioned LoadBalancer).
So, that end is sorted.

Also, after this PR (https://github.com/kyverno/kyverno/commit/f1491fe6d3e813682c4efaafd29e0f6a16e2f94a), the metrics will be exposed as an entirely different service with the name `kyverno-svc-metrics`. Hence, the user can choose to publicly expose metrics (By configuring the `kyverno-svc-metrics` to be of the type NodePort or LoadBalancer) and still keep the original `kyverno-svc` private to the cluster as it's not needed to be exposed.

yashvardhan-kukreja replied:
I'll update the docs accordingly with this :)

JimBugwadia reviewed (changes requested):
Please see prior comments to fix the titles and placement of the chapter.

yashvardhan-kukreja commented:
Hi @JimBugwadia , if possible, can you get this PR reviewed by today as tomorrow, CNCF is publishing the blog I wrote encapsulating my work with this project and I have linked these docs in the blog as a reference in multiple places.

JimBugwadia commented:
Will do. For future reviews, please "resolve comments" for all issues addressed and  re-request reviews, to indicate when changes are ready and help optimize the review process.

JimBugwadia commented on content/en/docs/Monitoring Kyverno With Prometheus Metrics/_index.md:
````diff
@@ -0,0 +1,74 @@
+...
+```
+
+## Metrics and a ready-to-use Grafana Dashboard
````
Change to "Metrics" and make "Grafana Dashboard" a separate section.

yashvardhan-kukreja replied:
tracking them in #184

JimBugwadia reviewed (approved):
Some minor changes.

- Make Grafana a separate top level title

We can shorten these titles as well:

Under "Metrics":

**Information around Policies and Rules" to **Policies and Rule Counts**

**Information around results of Rule/Policy executions** to **Policies and Rule Execution Counts**

**Information around Policy-level changes** to "Policy Change Counts"





I have approved the PR. We can merge and use a separate PR for the these changes.

JimBugwadia commented:
@yashvardhan-kukreja - see above. 

Let me know if you want to make the suggested changes with this PR or file a separate issue.

yashvardhan-kukreja commented:
Sure, Jim. Another PR sounds like a good idea.
Please merge this PR and I'll raise another PR addressing your suggestions.

Thanks :)

JimBugwadia commented:
@yashvardhan-kukreja - there is one change pending that you need to mark complete to allow the merge. It is currently blocked. Please see above.

JimBugwadia commented:
Also @yashvardhan-kukreja, please log a new issue to capture the pending changes (from comments) and link here. Thanks!

yashvardhan-kukreja commented:
> 
> Some minor changes.
> 
> - Make Grafana a separate top level title
> 
> We can shorten these titles as well:
> 
> Under "Metrics":
> 
> **Information around Policies and Rules" to **Policies and Rule Counts**
> 
> **Information around results of Rule/Policy executions** to **Policies and Rule Execution Counts**
> 
> **Information around Policy-level changes** to "Policy Change Counts"
> 
> 
> 
> 
> 
> I have approved the PR. We can merge and use a separate PR for the these changes.
> 

Are you talking about these changes Jim?

JimBugwadia commented:
There are two things you need to do:

1) Complete the changes previously requested by Chip, and then mark as complete in this PR to allow / unblock the merge. See PR details above and the screenshot:

![image](https://user-images.githubusercontent.com/5489990/121905853-689fab80-ccdf-11eb-9025-89b4d927cd3d.png)

2) File a separate issue for the changes I requested, which you said you will do in a separate PR and link that here.

yashvardhan-kukreja commented:
> There are two things you need to do:
> 
> 1. Complete the changes previously requested by Chip, and then mark as complete in this PR to allow / unblock the merge. See PR details above and the screenshot:
> 
> ![image](https://user-images.githubusercontent.com/5489990/121905853-689fab80-ccdf-11eb-9025-89b4d927cd3d.png)
> 
> 1. File a separate issue for the changes I requested, which you said you will do in a separate PR and link that here.

1. Jim I already addressed Chip's suggested changes in the further commits.
2. Added this issue - #184

yashvardhan-kukreja commented:
@chipzoller can you approve the PR for the changes you suggested to implement in the beginning? 

Thanks!

JimBugwadia commented:
@yashvardhan-kukreja  thanks for logging the new issue. Have you marked those changes as completed and re-requested the review, as previously suggested?

yashvardhan-kukreja commented:
Jim, I have re-requested the review but sorry I couldn't find the option for marking them as done and completed.

JimBugwadia commented:
Ok, thanks @yashvardhan-kukreja!

@chipzoller - when you get a chance, please update your review, if everything looks good to merge.
